src/components/forms/TextInput.jsx

```diff
@@ -0,0 +1,115 @@
+import React, { useEffect } from 'react';
+import PropTypes from 'prop-types';
+import { throttle, debounce } from 'throttle-debounce';
+import IconInput from 'components/forms/IconInput';
+import DecoratedInput from 'components/forms/DecoratedInput';
+
+export const throttleValue = 500;
+
+const TextInput = ({
+  label,
+  errorText,
+  throttleOnChange,
+  required,
+  onChange,
+  type,
+  IconLeft,
+  IconRight,
+  after,
+  before,
+  ...props
+}) => {
+  // Refer to the following for an explanation on
+  // the throttling / debouncing approach:
+  // https://www.peterbe.com/plog/how-to-throttle-and-debounce-an-autocomplete-input-in-react
+  let throttleOnChangeThrottled;
+  let throttleOnChangeDebounced;
+
+  useEffect(() => {
+    if (throttleOnChange) {
+      throttleOnChangeThrottled = throttle(throttleValue, throttleOnChange);
+      throttleOnChangeDebounced = debounce(throttleValue, throttleOnChange);
+    }
+  }, [throttleOnChange]);
+
+  const inputOnChange = (e) => {
+    e.persist();
+    onChange(e);
+    if (e.target.value.length < 5) {
+      throttleOnChangeThrottled(e);
+    } else {
+      throttleOnChangeDebounced(e);
+    }
+  };
+
+  const Element = label ? 'label' : 'div';
+
+  return (
+    <>
+      <Element className="fdsTextInput-root">
+        {label && (
+          <div className="fdsTextInput-label">
+            {label}
+            {required && <span className="color--red">&nbsp;*</span>}
+          </div>
+        )}
+        <DecoratedInput after={after} before={before}>
+          <IconInput IconLeft={IconLeft} IconRight={IconRight}>
+            <input
+              {...props}
+              onChange={inputOnChange}
+              type={type}
+              className="fdsTextInput"
+            />
+          </IconInput>
+        </DecoratedInput>
+      </Element>
+      {errorText && <div className="color--red type--data">{errorText}</div>}
+    </>
+  );
+};
+
+TextInput.defaultProps = {
+  type: 'text',
+  onChange: () => {},
+  throttleOnChange: () => {},
+};
+
+TextInput.propTypes = {
+  errorText: PropTypes.string,
+
+  /** Label used for `label` element */
+  label: PropTypes.string,
```

Comment:
How do you feel about making this required and having a `showLabel` prop the way we do with current checkbox?

If a developer wants to hide the label, we use the `label` string on an `aria-label` attribute of the input.

Reply:
So yeah I guess this becomes a bit tricky. I can see pros and cons of both sides of this.

If we force the developer to pass a label regardless, then what happens to the visual label that they need to show users? Do they still use their own label (so that there are 2 labels)? Or do they just add a span with some text, and the "true" label is on the `aria-label`? I sort of see the benefit of making label `required` so that devs don't abuse not passing a label (which will 100% happen), but then again I sort of question the developer having to apply "2" labels, the one they pass to the component, and then the one they specify themselves, and then the whole thing where they click on their own label... would that focus the input. Lot of... variables here 🙂

Reply:
The pattern would be...

```
label: PropTypes.string.isRequired,
showLabel: PropTypes.bool, // default `true` - if set to `false`, apply `label` to `aria-label`
```

Reply:
right but I don't get the surrounding things that happen as a result of this. We should maybe chat about it, it's slightly confusing.